src/devices/nm-device.c

```diff
@@ -2157,17 +2170,213 @@ nm_device_get_physical_port_id (NMDevice *self)
 
 /*****************************************************************************/
 
+typedef enum {
+	CONCHECK_SCHEDULE_UPDATE_INTERVAL,
+	CONCHECK_SCHEDULE_CHECK_EXTERNAL,
+	CONCHECK_SCHEDULE_CHECK_PERIODIC,
+	CONCHECK_SCHEDULE_RETURNED_MIN,
+	CONCHECK_SCHEDULE_RETURNED_BUMP,
+	CONCHECK_SCHEDULE_RETURNED_MAX,
+} ConcheckScheduleMode;
+
+static NMDeviceConnectivityHandle *concheck_start (NMDevice *self,
+                                                   NMDeviceConnectivityCallback callback,
+                                                   gpointer user_data,
+                                                   gboolean is_periodic);
+
+static void concheck_periodic_schedule_set (NMDevice *self,
+                                            ConcheckScheduleMode mode);
+
+static gboolean
+concheck_periodic_timeout_cb (gpointer user_data)
+{
+	NMDevice *self = user_data;
+
+	concheck_periodic_schedule_set (self, CONCHECK_SCHEDULE_CHECK_PERIODIC);
+	concheck_start (self, NULL, NULL, TRUE);
+	return G_SOURCE_CONTINUE;
+}
+
+static gboolean
+concheck_periodic_schedule_do (NMDevice *self, gint64 interval_ns)
+{
+	NMDevicePrivate *priv = NM_DEVICE_GET_PRIVATE (self);
+	gboolean periodic_check_disabled = FALSE;
+
+	/* we always cancel whatever was pending. */
+	if (nm_clear_g_source (&priv->concheck_p_cur_id))
+		periodic_check_disabled = TRUE;
+
+	if (priv->concheck_p_max_interval == 0) {
+		/* periodic checks are disabled */
+		goto out;
+	}
+
+	nm_assert (interval_ns >= 0);
+
+	if (!nm_device_is_real (self))
+		goto out;
+
+	/* we enable periodic checks for every device state (except UNKNOWN). Especially with
+	 * unmanaged devices, it is interesting to know whether we have connectivity on that device. */
```

Comment:
We should at least check that the device is realized, and probably also avoid loopback.

I wonder if we should keep ignoring unmanaged devices as before, because the promise is that they don't exist from
NM point of view. But I agree that if there is a unmanaged device that can reach the internet, it's useful to know it to show the real connectivity state.

Reply:
> We should at least check that the device is realized,

a few lines above there is a check for `nm_device_is_real()`. Added check for loopback.

Let's enable connectivity check eagerly for the moment, and see how that works. I think there is a point in doing connectivity checks for unmanaged devices.


Repushed.

Reply:
> a few lines above there is a check for nm_device_is_real(). Added check for loopback.

Right, it's the manual connectivity check that doesn't verify whether a device is realized and so we start the check also on devices that don't exist:

    connectivity: (pppoe) start request to 'http://fedoraproject.org/static/hotspot.txt'
    connectivity: (pppoe) check completed: LIMITED; check failed with curl status 45

Reply:
Ah... the idea was that `nm_device_check_connectivity()` never fails and also doesn't schedule it's own idle handler -- because `nm_connectivity_check_start()` already schedules an idle handler when there is nothing to do. The point was a consistent behavior. You are right, that is not great. Pushed a fixup. How about now?

Reply:
LGTM